org.jacoco.core/src/org/jacoco/core/runtime/OfflineInstrumentationCompanionAccessGenerator.java

```diff
@@ -0,0 +1,118 @@
+import java.util.UUID;
+
+import org.jacoco.core.JaCoCo;
+import org.jacoco.core.internal.instr.Companions;
+import org.jacoco.core.internal.instr.InstrSupport;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+
+/**
+ * Strategy that creates "companion" class in "offline" mode, which will hold
+ * state of instrumented classes. This allows to avoid addition of members to
+ * classes.
+ */
+public class OfflineInstrumentationCompanionAccessGenerator
+		implements IExecutionDataAccessorGenerator {
+
+	private final OfflineInstrumentationAccessGenerator generator;
+
+	private final String companionName;
+
+	private final Set<String> instrumented = new HashSet<String>();
+
+	private final ClassWriter cw;
+
+	private final MethodVisitor mv;
+
+	/**
+	 * Creates a new instance for offline instrumentation.
+	 */
+	public OfflineInstrumentationCompanionAccessGenerator() {
+		this(JaCoCo.RUNTIMEPACKAGE.replace('.', '/') + "/Offline");
+	}
+
+	OfflineInstrumentationCompanionAccessGenerator(
+			final String runtimeClassName) {
+		this(runtimeClassName, Companions.COMPANION_NAME
+				+ UUID.randomUUID().toString().replace('-', '_'));
+	}
+
+	OfflineInstrumentationCompanionAccessGenerator(
+			final String runtimeClassName, final String companionName) {
+		this.generator = new OfflineInstrumentationAccessGenerator(
+				runtimeClassName);
+		this.companionName = companionName;
+
+		cw = new ClassWriter(0);
+		cw.visit(Opcodes.V1_1,
+				Opcodes.ACC_SYNTHETIC | Opcodes.ACC_PUBLIC | Opcodes.ACC_FINAL,
+				companionName, null, "java/lang/Object", null);
+		mv = cw.visitMethod(Opcodes.ACC_STATIC, "<clinit>", "()V", null, null);
+		mv.visitCode();
+	}
+
+	public int generateDataAccessor(final long classid, final String classname,
+			final int probecount, final MethodVisitor mv) {
+		if (instrumented.size() == Companions.FIELDS_PER_CLASS) {
+			throw new UnsupportedOperationException();
+		}
+
+		final String fieldName = "p" + Long.toString(classid);
+
+		if (instrumented.add(classname)) {
+			cw.visitField(
+					Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC
+							| Opcodes.ACC_TRANSIENT | Opcodes.ACC_FINAL,
+					fieldName, InstrSupport.DATAFIELD_DESC, null, null);
+
+			generator.generateDataAccessor(classid, classname, probecount,
+					this.mv);
+
+			this.mv.visitFieldInsn(Opcodes.PUTSTATIC, companionName, fieldName,
+					InstrSupport.DATAFIELD_DESC);
+		}
+
+		mv.visitFieldInsn(Opcodes.GETSTATIC, companionName, fieldName,
+				InstrSupport.DATAFIELD_DESC);
+		return 1;
+	}
+
+	/**
+	 * Returns the name of generated "companion" class.
+	 *
+	 * @return name of generated "companion" class
+	 */
+	public String getClassName() {
+		return companionName;
+	}
+
+	/**
+	 * Returns the bytecode of generated "companion" class.
+	 *
+	 * @return bytecode of generated "companion" class
+	 */
+	public byte[] getClassDefinition() {
+		mv.visitInsn(Opcodes.RETURN);
+		mv.visitMaxs(4, 0);
```

Comment:
Duplicating the logic seems worse to me.  If you don't want the complexity of using the return value from generateDataAccessor, then perhaps add a MAX_STACK constant to OfflineInstrumentationAccessGenerator and use it in both places?